### PR TITLE
Retry upgrades even if values or version are unchanged and add releasemaxfailed resource

### PR DESCRIPTION
**Pull request by rossf7**
Towards https://github.com/giantswarm/giantswarm/issues/17622

## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: service/controller/chart/resource/namespace/resource.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ type Resource struct {
 	k8sWaitTimeout time.Duration
 }
 
-// New creates a new configured tcnamespace resource.
+// New creates a new configured namespace resource.
```
Comment (rossf7) [inline service/controller/chart/resource/namespace/resource.go:37]: Just fixing this.

File: service/controller/chart/resource/release/resource.go [hunk 2]
```diff
@@ -230,34 +228,6 @@ func (r *Resource) addHashAnnotation(ctx context.Context, cr v1alpha1.Chart, rel
...
-
-	for i := 0; i < project.ReleaseFailedMaxAttempts; i++ {
-		if history[i].Status != helmclient.StatusFailed {
-			return false, nil
-		}
-	}
-
-	// All failed so we exceeded the max attempts.
-	return true, nil
-}
```
Comment (rossf7) [inline service/controller/chart/resource/release/resource.go:259]: This moves to the new releasemaxhistory resource.

File: service/controller/chart/resource/release/update.go [hunk 3]
```diff
@@ -254,28 +254,17 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
...
-	if isReleaseFailed(currentReleaseState, desiredReleaseState) {
-		r.logger.Debugf(ctx, "the %#q release is in status %#q and cannot be updated", desiredReleaseState.Name, currentReleaseState.Status)
+	// We check the controller context and if the release has failed more
+	// than the max number of attempts we stop updating. Otherwise too many
+	// release secrets will be created.
+	//
+	// The releasemaxhistory resource handles deleting the oldest secret if
+	// it is over 1 minute old. So we still retry but at a slower rate.
+	if currentReleaseState.Status == helmclient.StatusFailed && cc.Status.Release.FailedMaxAttempts {
+		r.logger.Debugf(ctx, "the %#q release is in status %#q and has failed %d times", desiredReleaseState.Name, currentReleaseState.Status, project.ReleaseFailedMaxAttempts)
```
Comment (rossf7) [inline service/controller/chart/resource/release/update.go:264]: We should retry even if values or version are unchanged for situations like failed webhooks.

This should mean we get less pages but we also slow the retries and still ensure that too many release secrets are not created.

File: service/controller/chart/resources.go [hunk 4]
```diff
@@ -135,9 +151,15 @@ func newChartResources(config chartResourcesConfig) ([]resource.Interface, error
...
 	resources := []resource.Interface{
+		// tiller migration deletes tiller once helm 3 migration is done.
 		tillerMigrationResource,
+		// namespace creates the release namespace and allows setting metadata.
 		namespaceResource,
+		// release max history ensures not too many helm release secrets are created.
+		releaseMaxHistoryResource,
+		// release manages Helm releases and is the most important resource.
 		releaseResource,
+		// status resource manages the chart CR status.
```
Comment (rossf7) [inline service/controller/chart/resources.go:162]: We have these comments in other operators and I think are useful here too. 🤞 we can delete tiller migration soon.

**Review by rossf7 — COMMENTED** [review]
Tested on gorgoth

File: service/controller/chart/resource/releasemaxhistory/create.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+	if len(secrets.Items) != 1 {
+		return false, microerror.Maskf(executionFailedError, "expected 1 release secret got %d", len(secrets.Items))
+	}
+
+	secret := secrets.Items[0]
+	diff := time.Since(secret.CreationTimestamp.Time)
+	if diff.Minutes() < 1 {
```
Comment (rossf7) [inline service/controller/chart/resource/releasemaxhistory/create.go:94]: I think 1 minute is enough to ensure upgrades are rate limited. But maybe longer is needed. @tomahawk28 WDYT?

Comment (tomahawk28) [reply service/controller/chart/resource/releasemaxhistory/create.go:94]: Sure, 1 minute is reasonable to wait.

File: service/controller/chart/resource/releasemaxhistory/create.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+		r.logger.Debugf(ctx, "revision %d for release %#q is < 1 minutes old", rev.Revision, releaseName)
+		r.logger.Debugf(ctx, "canceling resource")
```
Comment (tomahawk28) [inline service/controller/chart/resource/releasemaxhistory/create.go:96]: I think we are not canceling the resource at this moment, just returning `false`.

Comment (rossf7) [reply service/controller/chart/resource/releasemaxhistory/create.go]: Good catch, thx! No enough ☕ when I did this

**Comment by rossf7** [comment]
I did some more testing with workload clusters and looks fine.

@tomahawk28 PTAL

**Review by tomahawk28 — APPROVED** [review]
LGTM!

**Pull request merged**